lib/THC/THCTensorMode.cuh

```diff
@@ -0,0 +1,282 @@
+  }
+  __syncthreads(); // barrier for bmem initialization
+
+  // Next, we initialize another shared memory buffer cmem which will be used in the
+  // Segmented Prefix Sum
+  __shared__ unsigned int  cmem[Power2Size];
+
+  // We set cmem to be the negation of bmem. In particular, we can think of cmem[i] = true
+  // iff A[i-1] == A[i] in our original sorted slice.
+  cmem[tidx] = !bmem[tidx];
+  cmem[stidx] = !bmem[stidx];
+  __syncthreads(); // barrier for cmem initialization
+
+  // Next, we perform a segmented prefix sum on the neighboring elements, where
+  // the presence of a one indicates the start of a segment. In this case bmem acts
+  // as the segment start flags, and cmem is the buffer to be summed:
+  //
+  // Input  (cmem)  = [0, 1, 0, 1, 0, 1, 1, 0, 0, 0, 1, 0, 0]
+  // Flag   (bmem)  = [1, 0, 1, 0, 1, 0, 0, 1, 1, 1, 0, 1, 1]
+  // Output (cmem)  = [0, 1, 0, 1, 0, 1, 2, 0, 0, 0, 1, 0, 0]
+  //
+  // Afterwards, the cmem buffer contains the lengths of the segments (minus 1), i.e. the counts
+  // of each element in the original input.
+  segmentedInclusivePrefixScan<unsigned int, BinaryAddOp<unsigned int>, Power2Size>(cmem, bmem, BinaryAddOp<unsigned int>());
+
+  // Our last shared memory buffer is used to track indices
+  __shared__ unsigned int imem[Power2Size];
+
+  // initialize the indices buffer such that imem[i] = i
+  imem[tidx] = tidx;
+  imem[stidx] = stidx;
+  __syncthreads(); // barrier for both the scan and the imem initialization
+
+  // At this point, we need to find the maximum element in the cmem buffer.
+  // This element will represent the count (-1) of the mode. Because of the
+  // way we have set up the problem, the index where this mode occurs will
+  // also be the location of the mode value in the sorted array, e.g.
+  //
+  // smem = [0, 0, 1, 1, 1, 2]
+  // cmem = [0, 1, 0, 1, 2, 0]
+  //                     ^
+  //                     maximum value, also aligned with mode = 1
+  //
+  // We perform a block wide max-reduction of the cmem buffer, and bring imem
+  // along with it.
+  //
+  // Loop 1 (Power2Size = offset = 4):
+  //
+  // (0, 4) --> cmem[4] = 2 > cmem[0] = 0, so update cmem[0], imem[0]
+  // (1, 5) --> cmem[5] = 0 <= cmem[1] = 1, do nothing
+  //
+  // Now:         0  1  2  3  4  5
+  //      cmem = [2, 1, 0, 1, 2, 0]
+  //      imem = [4, 1, 2, 3, 4, 5]
+  //
+  // Loop 2 (offset = 2)
+  //
+  // (0, 2) --> cmem[2] == 0 <= cmem[0] = 2, do nothing
+  // (1, 3) --> cmem[3] == 1 <= cmem[1] = 1, do nothing
+  //
+  // Now:         0  1  2  3  4  5
+  //      cmem = [2, 1, 2, 1, 2, 0]
+  //      imem = [4, 1, 4, 3, 4, 5]
+  //
+  // Loop 3 (offset = 1)
+  //
+  // (0, 1) --> cmem[1] == 1 <= cmem[0] = 2, do nothing
+  //
+  // So at the end at cmem[0] we have the maximum count = 2, and the
+  // corresponding index = 4
+#pragma unroll
+  for (unsigned int offset = Power2Size / 2; offset > 0; offset >>= 1) {
```

Comment:
This is more duplicated reduction code, would be nice to restructure to use helper functions like `reduceBlock`.

The issue is that you have different buffers declared for using shared memory, but in this case you are reducing what is effectively a:

```
struct ModeCount { int count; int index; };
```

with maximum count. 

`smem` is used with `bmem`, and `cmem` depends upon `bmem`. 
`imem` is used together with `cmem`.

All of them do not need to be resident at the same point in time. This is decreasing occupancy as well. In fact, at most you need (max(sizeof(T), sizeof(unsigned int)) + sizeof(unsigned int)) * Power2Size shared memory. You can dynamically allocate this instead, and repurpose the memory at each step.

Then, you can view that memory as an array of my `struct ModeCount` above, and just use `reduceBlock` on that.